### DISCUSSION TITLE
Docker and IPV6

### DISCUSSION THREAD

**Discussion by lorenzo-gomez-windhover** [ControlSystemStudio/phoebus#2156] [General]
I have been recently invested some effort in creating a container for phoebus. The motivation for this is so that we as devs can have a consistent environment and is easy to reproduce when it comes to testing and deployment. We could also use the container in our CI. Everything works great except for the fact that the following test fails:
`org.epics.pva.common.NetworkTest`

This is because Docker does not support IPV6 out of the box. As per [documentation](https://docs.docker.com/config/daemon/ipv6/), one can configure docker(from the host) to support it. However, even if we managed to do this in CI, GitHub Actions does not support IPV6 at the moment. You can read more about that [here](https://github.com/actions/virtual-environments/issues/668).

I'm opening a new discussion on this in an effort to collaborate and share knowledge on this since, if we would like containers, there are multiple steps involved to get there. Hopefully the GithubActions infra can improve over time and we can tackle this issue.

**Comment by kasemir** [comment]
Well, bummer.
For some EPICS sites there's a push to enable IPv6 on all new installations and phase out IPv4.
So we're preparing PVA and maybe also CA to start supporting IPv6, but of course IPv4 support will remain in place for a long time.

Is your docker issue limited to just the associated tests failing? That would of course be expected on a platform that doesn't support IPv6. We could try to come up with a way to disable just the IPv6 tests, for example by setting some environment variable. Can you just build without the tests (`mvn -DskipTests clean install`), and does the result then work alright?

**Comment by shroffk** [comment]
Hi @lorenzo-gomez-windhover 

Thanks for looking into this... there seem to be a lot of people/projects requesting this https://github.com/actions/virtual-environments/issues/668 but it still seems to be marked won't fix.. so I don't think I would hold my breath for this :(

First, I think we should still go ahead and try to get a IPv6 enabled docker image.. at least for local testing. I am hoping @andersharrisson-ess can help since they have already worked on this.

For the github Actions, like Kay suggested, I can add a profile which specifically ignore the Network test. This was we can still run all the other unit tests and the new UI tests on CI.

**Reply by lorenzo-gomez-windhover** [reply]
> For the github Actions, like Kay suggested, I can add a profile which specifically ignore the Network test. This was we can still run all the other unit tests and the new UI tests on CI.

Yes, I had to do this as well. If you would like, I could probably do this in a PR as well as update the config of CI to use the docker image.

**Reply by andersharrisson-ess** [reply]
> First, I think we should still go ahead and try to get a IPv6 enabled docker image.. at least for local testing. I am hoping @andersharrisson-ess can help since they have already worked on this.

As @georgweiss said, we have configured Docker to support IPv6 at ESS. Since we are building our Phoebus releases on our mirrored repository in GitLab, we have IPv6 support on the runners.
I think you need to push for https://github.com/actions/virtual-environments/issues/668 to be able to run the full test suite on GitHub Actions, even if I agree that it seems unlikely that it will be fixed in the near future.

**Comment by georgweiss** [comment]
At ESS we're running builds in Gitlab CI, which obviously can be configured to support IPV6 inside Docker.

**Comment by kasemir** [comment]
Where exactly does the `org.epics.pva.common.NetworkTest` fail? Maybe the test can self-determine if IPv6 is supported or not...

**Reply by lorenzo-gomez-windhover** [reply]
If I comment out the following, then all tests pass with `mvn tests` inside of the container.
```
--- a/core/pva/src/test/java/org/epics/pva/common/NetworkTest.java
+++ b/core/pva/src/test/java/org/epics/pva/common/NetworkTest.java
@@ -112,7 +112,7 @@ public class NetworkTest
                                           .map(iface -> iface.getHostAddress())
                                           .collect(Collectors.joining(", "));
             System.out.println("Interface addresses: " + iface_addr);
-            assertTrue(iface_addr.contains("0:0:0:0:0:0:0:1"));
+            //assertTrue(iface_addr.contains("0:0:0:0:0:0:0:1"));
             assertFalse(addr.isBroadcast());
         }
     }
@@ -125,17 +125,17 @@ public class NetworkTest
         final List<AddressInfo> infos = Network.parseAddresses(spec, PVASettings.EPICS_PVA_BROADCAST_PORT);
         for (AddressInfo info : infos)
             System.out.println(info);
-        assertEquals(4, infos.size());
+        //assertEquals(4, infos.size());
 
         assertTrue(infos.get(0).getAddress().getAddress() instanceof Inet4Address);
         assertTrue(infos.get(1).getAddress().getAddress() instanceof Inet6Address);
         assertTrue(infos.get(2).getAddress().getAddress() instanceof Inet4Address);
-        assertTrue(infos.get(3).getAddress().getAddress() instanceof Inet6Address);
+        //assertTrue(infos.get(3).getAddress().getAddress() instanceof Inet6Address);
 
         assertTrue(((Inet4Address)infos.get(2).getAddress().getAddress()).isMulticastAddress());
-        assertTrue(((Inet6Address)infos.get(3).getAddress().getAddress()).isMulticastAddress());
+        //assertTrue(((Inet6Address)infos.get(3).getAddress().getAddress()).isMulticastAddress());
 
         assertTrue(infos.get(2).getInterface() != null);
-        assertTrue(infos.get(3).getInterface() != null);
+        //assertTrue(infos.get(3).getInterface() != null);
     }
 }

```

**Comment by kasemir** [comment]
Thanks for identifying the lines that fail. Makes sense now since you won't have a local "::1" address.
In https://github.com/ControlSystemStudio/phoebus/pull/2160 I've added support for a property `ignore_local_ipv6`.

If you somehow pass `-Dignore_local_ipv6=true` to the build, be it via JDK_JAVA_OPTIONS or maven profile or ..., it should skip those lines.

**Reply by lorenzo-gomez-windhover** [reply]
Thanks! Will be sure to make that PR as soon as possible.
